### PR TITLE
Add NTP protocol analyzer

### PR DESCRIPTION
**Pull request by mauropalumbo75**
I wrote an NTP protocol analyzer (starting from some work done by Vlad Grigorescu).

The analyzer parses almost every field required by the protocol and writes some relevant information in a new log file called "ntp.log". I would welcome comments if anyone believe some more data would be useful in the log file or if there is any suggestion how to improve the analyzer.

I added some pcap files and tests too.

### CODE REVIEW

File: src/analyzer/protocol/ntp/NTP.cc [hunk 1]
```diff
@@ -1,114 +1,38 @@
...
-	// to convert an old-style UDP analyzer.
-	if ( is_orig )
-		Request(data, len);
-	else
-		Reply(data, len);
+	delete interp;
 	}
 
-int NTP_Analyzer::Request(const u_char* data, int len)
+void NTP_Analyzer::Done()
```
Comment (timwoj) [inline src/analyzer/protocol/ntp/NTP.cc]: If all you're doing is calling the parent method in this method, you don't need to override it. The child version can be removed.

File: scripts/base/protocols/ntp/main.zeek [hunk 2]
```diff
@@ -0,0 +1,209 @@
+module NTP;
```
Comment (timwoj) [inline scripts/base/protocols/ntp/main.zeek]: This file in general needs a little work on formatting. The Zeek repo uses Whitesmiths style for all code, including zeek scripts. There's a number of tabs/spaces issues as well (prefer tabs).

File: src/analyzer/protocol/ntp/events.bif [hunk 3]
```diff
@@ -4,18 +4,12 @@
 ## See `Wikipedia <http://en.wikipedia.org/wiki/Network_Time_Protocol>`__ for
 ## more information about the NTP protocol.
 ##
-## u: The connection record describing the corresponding UDP flow.
+## c: The connection record describing the corresponding UDP flow.
 ##
-## msg: The parsed NTP message.
+## is_orig:
```
Comment (timwoj) [inline src/analyzer/protocol/ntp/events.bif]: Missing documentation?

File: src/analyzer/protocol/ntp/ntp-protocol.pac [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+# As in RFC 1119
+type NTP_CONTROL_MAC = record {
+        key_id		: uint32;
+     	crypto_checksum	: bytestring &length=8;
+} &length=12;
+
+# As defined in RFC 5906
+type Extension_Field = record {
+  first_byte_ext: uint8;
```
Comment (timwoj) [inline src/analyzer/protocol/ntp/ntp-protocol.pac]: Indentation here and below should be tabs.

File: scripts/base/protocols/ntp/dpd.sig [hunk 5]
```diff
@@ -0,0 +1,12 @@
+signature dpd_ntp {
+
+ 	ip-proto == udp
+
+
+ 	# ## TODO: Define the payload. When Bro sees this regex, on
```
[the remark this answers is not in the document]
Comment (mauropalumbo75) [reply scripts/base/protocols/ntp/dpd.sig]: I did not use a signature for this analyzer. So, I guess it is better to remove the TODOs here and in the following questions as well as the entire file dpd.sig. 
In the NTP protocol at least on of the two ports (orig or resp) must be 123. Would signature detection be so important here?

File: scripts/base/protocols/ntp/__load__.zeek [hunk 6]
```diff
@@ -0,0 +1,2 @@
+@load ./main
+#@load-sigs ./dpd.sig
```
[the remark this answers is not in the document]
Comment (mauropalumbo75) [reply scripts/base/protocols/ntp/__load__.zeek]: see above reply

File: scripts/base/protocols/ntp/main.zeek [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+# TODO: The recommended method to do dynamic protocol detection
```
[the remark this answers is not in the document]
Comment (mauropalumbo75) [reply scripts/base/protocols/ntp/main.zeek]: see above reply

File: scripts/base/protocols/ntp/main.zeek [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        	## An integer specifying the command function. Values currently defined includes:
+        	## 1 read status command/response
+        	## 2 read variables command/response
+        	## 3 write variables command/response
+        	## 4 read clock variables command/response
+        	## 5 write clock variables command/response
+        	## 6 set trap address/port command/response
+        	## 7 trap response
+        	## Other values are reserved.
+        	OpCode          : count &log;
```
[the remark this answers is not in the document]
Comment (mauropalumbo75) [reply scripts/base/protocols/ntp/main.zeek]: probably better as you say

File: scripts/base/protocols/ntp/dpd.sig (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsiwek) [reply scripts/base/protocols/ntp/dpd.sig]: Just removing unnecessary comments/TODOs sounds fine to me in this case.

File: src/analyzer/protocol/ntp/NTP.cc (hunk 1 above)
Comment (mauropalumbo75) [reply src/analyzer/protocol/ntp/NTP.cc]: correct. I removed the method

File: src/analyzer/protocol/ntp/events.bif (hunk 3 above)
Comment (mauropalumbo75) [reply src/analyzer/protocol/ntp/events.bif]: yes. corrected

File: scripts/base/protocols/ntp/main.zeek (hunk 2 above)
Comment (mauropalumbo75) [reply scripts/base/protocols/ntp/main.zeek]: ok, done

File: src/analyzer/protocol/ntp/ntp-protocol.pac (hunk 4 above)
Comment (mauropalumbo75) [reply src/analyzer/protocol/ntp/ntp-protocol.pac]: ok

**Comment by jsiwek** [comment]
Thanks for the PR, it's merged w/ some changes:

* Formatting/style (still a bunch of tab/whitespace misuse, please make it more consistent in initial PRs to save us some time)
* Renamed the new NTP record types for consistency/convention
* Got rid of the separate/unused IPv6 reference ID field
* Tried to get the default `ntp.log` under control and smaller:

  * Combined the reference ID into a single field
  * Wasn't sure how useful the key ID and message digest are, so removed
  * Shoving fields that are exclusive to mode 6 and 7 into the same log as everything else seems like it will end up populating it with a bunch of nil fields most of the time, which is not useful, so I removed those.  May be better to separate those out into individual log files, one for each, or to try to improve formatting to make it understandable to human reader: e.g. translate op/request codes or match up any request/response transactions.

@sethhall @grigorescu any comments on the sanity of the default `ntp.log` ?

**Comment by grigorescu** [comment]
Wow, that was quick!

There are some fields that feel a bit extraneous, like key_id and digest. 

I'm wondering if the log should be split in 2, ntp and ntp_control or some-such. modes 1-5 are really for time synchronization, while modes 6 and 7 are more for managing the ntp server(s) themselves (or for DDoS attacks). Looking at the baselines, it looks like we either have the time synchronization fields set, or we have the control fields set, but not both (which makes sense).

I'll see if I can get it running on live traffic to get a better feel for it.

**Comment by mauropalumbo75** [comment]
I expect most NTP  messages to be "regular" mode 1-5 ones, as Vlad points out the other modes are for control, debugging, special purposes. Since an NTP packet can either be a "regular" mode 1-5 one or a "special" mode 6-7, the fields in the ntp.log are never completely filled and it probably makes sense to split the log file into two separate ones. But as mode 6-7 messages can be used maliciously, I believe it is better to log this info.

Honestly, I am not sure how useful are key_id and digest in the log, but I would live them in the event...

**Comment by jsiwek** [comment]
@grigorescu @mauropalumbo75 thanks for the comments, think we arrived at similar conclusions per my earlier comments about what I changed during merge:

* I kept `key_id` and `digest` available in the event, but they aren't in `ntp.log` by default (if anyone thinks we can trim further, just say so since I don't have a good feel for that)
* I removed mode 6-7 messages from default `ntp.log`

If one wants a separate log for mode 6-7, it's still possible for them to make that themselves.  If it's common to want that info by default, submit a new PR that does that.

**Pull request merged**